File: UEReflectWatch/Projectresolver.cs
using System.IO;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace UEReflectWatch
{
    public sealed class UnrealProject
    {
        public string UprojectPath { get; }
        public string ProjectName { get; }
        public string EnginePath { get; }
        public string EngineVersion { get; }

        public UnrealProject(string uprojectPath, string projectName, string enginePath, string engineVersion)
        {
            UprojectPath = uprojectPath;
            ProjectName = projectName;
            EnginePath = enginePath;
            EngineVersion = engineVersion;
        }

        public string BuildBatPath =>
            Path.Combine(EnginePath, "Engine", "Build", "BatchFiles", "Build.bat");

        public string EditorExePath =>
            Path.Combine(EnginePath, "Engine", "Binaries", "Win64", "UnrealEditor.exe");
    }

    internal sealed class UprojectJson
    {
        [JsonPropertyName("EngineAssociation")]
        public string? EngineAssociation { get; set; }
    }

    public static class ProjectResolver
    {
        // Default install root used by the Epic Games Launcher on Windows.
        private const string EpicDefaultRoot = @"C:\Program Files\Epic Games";

        public static UnrealProject? Resolve(string solutionDirectory, string? enginePathOverride = null)
        {
            var uprojectPath = FindUproject(solutionDirectory);
            if (uprojectPath is null) return null;

            return ResolveFromUproject(uprojectPath, enginePathOverride);
        }

        private static string? FindUproject(string dir)
        {
            try
            {
                foreach (var file in Directory.GetFiles(dir, "*.uproject", SearchOption.TopDirectoryOnly))
                    return file;
            }
            catch { }
            return null;
        }

        private static UnrealProject? ResolveFromUproject(string uprojectPath, string? enginePathOverride)
        {
            try
            {
                var raw = File.ReadAllText(uprojectPath);
                var json = JsonSerializer.Deserialize<UprojectJson>(raw);

                var projectName = Path.GetFileNameWithoutExtension(uprojectPath);
                var engineVersion = json?.EngineAssociation ?? string.Empty;

                // User override takes priority.
                if (!string.IsNullOrWhiteSpace(enginePathOverride) && Directory.Exists(enginePathOverride))
                {
                    return new UnrealProject(uprojectPath, projectName, enginePathOverride, engineVersion);
                }

                // Construct the default Epic Games Launcher install path directly
                // from the EngineAssociation version string in the .uproject file.
                // e.g. "5.7" -> C:\Program Files\Epic Games\UE_5.7
                if (string.IsNullOrWhiteSpace(engineVersion))
                    return null;

                var defaultPath = Path.Combine(EpicDefaultRoot, $"UE_{engineVersion}");
                if (Directory.Exists(defaultPath))
                    return new UnrealProject(uprojectPath, projectName, defaultPath, engineVersion);

                return null;
            }
            catch
            {
                return null;
            }
        }
    }
}

File: RiderPlugin/src/main/kotlin/com/ueReflectWatch/resolver/ProjectResolver.kt
package com.ueReflectWatch.resolver

import com.ueReflectWatch.runner.PlatformAdapter
import com.ueReflectWatch.settings.UEReflectWatchSettings
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File

data class UnrealProject(
    val uprojectPath: String,
    val projectName: String,
    val enginePath: String,
    val engineVersion: String,
    val adapter: com.ueReflectWatch.runner.PlatformAdapter
) {
    val buildScriptPath: String get() = adapter.buildScriptPath(enginePath)
    val editorExecutablePath: String get() = adapter.editorExecutablePath(enginePath)
    val buildPlatformArg: String get() = adapter.buildPlatformArg()
}

@Serializable
private data class UprojectJson(
    val EngineAssociation: String = ""
)

object ProjectResolver {

    private val json = Json { ignoreUnknownKeys = true }

    fun resolve(solutionDir: String, adapter: com.ueReflectWatch.runner.PlatformAdapter): UnrealProject? {
        val uprojectFile = findUproject(solutionDir) ?: return null
        return resolveFromUproject(uprojectFile, adapter)
    }

    private fun findUproject(dir: String): File? {
        return File(dir).listFiles()
            ?.firstOrNull { it.extension == "uproject" }
    }

    private fun resolveFromUproject(
        uprojectFile: File,
        adapter: com.ueReflectWatch.runner.PlatformAdapter
    ): UnrealProject? {
        return try {
            val raw = uprojectFile.readText()
            val parsed = json.decodeFromString<UprojectJson>(raw)
            val projectName = uprojectFile.nameWithoutExtension
            val engineVersion = parsed.EngineAssociation

            // User override takes priority.
            val settings = UEReflectWatchSettings.instance
            val overridePath = settings.enginePathOverride
            if (overridePath.isNotBlank() && File(overridePath).exists()) {
                return UnrealProject(
                    uprojectPath = uprojectFile.absolutePath,
                    projectName = projectName,
                    enginePath = overridePath,
                    engineVersion = engineVersion,
                    adapter = adapter
                )
            }

            if (engineVersion.isBlank()) return null

            val defaultPath = adapter.defaultEnginePath(engineVersion)
            if (!File(defaultPath).exists()) return null

            UnrealProject(
                uprojectPath = uprojectFile.absolutePath,
                projectName = projectName,
                enginePath = defaultPath,
                engineVersion = engineVersion,
                adapter = adapter
            )
        } catch (_: Exception) {
            null
        }
    }
}
